bazel/toolchain: fix llvm 18 toolcahin

rockwotj opened the pull request:
Recompile the llvm-18 toolchain using and old debian container and the exact same cmake options as the vtools build. This finally fixes the sanitizer build in Bazel. We also add back in clang-19, I think we could switch to clang-19, but I'll wait to do that for another PR.

We can incrementally add more LLVM options/optimizations as time goes on to the build, but starting with something that works and allows us to enable the sanitizers in the Bazel build is a better starting point.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
bazel/toolchain: add llvm 20 toolchain

For local testing, have not tried to build with it yet.

rockwotj committed:
base/units: fix clang-20 warnings

Fixes the following errors

```
src/v/base/units.h:27:29: error: identifier '_TiB' preceded by whitespace in a literal operator declaration is deprecated [-Werror,-Wdeprecated-literal-operator]
```

rockwotj committed:
iceberg/avro: fix clang-20 error

```

src/v/iceberg/tests/schema_avro_test.cc:26:5: error: no matching function for call to 'CmpHelperSTREQ'
   26 |     ASSERT_STREQ(parent->nameAt(pos).c_str(), expected_name.cbegin());
external/googletest~/googletest/include/gtest/gtest.h:1487:28: note: candidate function not viable: no known conversion from 'const_iterator' (aka '__wrap_iter<const char *>') to 'const char *' for 4th argument
```

rockwotj committed:
security: make regex work with libc++20

rockwotj committed:
delta_for: fixes for libc++20

Use raw pointer instead of iterator. The iterator is a different type
than a pointer now.

rockwotj committed:
compression: fixes for libc++20

rockwotj committed:
iceberg/json: fixes for libc++20

rockwotj committed:
config: fix for libc++20

rockwotj commented:
Hrm currently stuck on this and I don't understand the error...

```
ERROR: /home/rockwood/code/redpanda/src/v/crash_tracker/BUILD:3:20: Compiling src/v/crash_tracker/recorder.cc failed: (Exit 1): cc_wrapper.sh failed: error executing CppCompile command (from target //src/v/cra
sh_tracker:crash_tracker) external/toolchains_llvm~~llvm~llvm_20_toolchain/bin/cc_wrapper.sh -U_FORTIFY_SOURCE '--target=x86_64-unknown-linux-gnu' '-march=westmere' -U_FORTIFY_SOURCE -fstack-protector -fno-omi
t-frame-pointer ... (remaining 717 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from src/v/crash_tracker/recorder.cc:12:
In file included from bazel-out/k8-fastbuild/bin/src/v/crash_tracker/_virtual_includes/crash_tracker/crash_tracker/recorder.h:15:
In file included from bazel-out/k8-fastbuild/bin/src/v/crash_tracker/_virtual_includes/crash_tracker/crash_tracker/prepared_writer.h:15:
In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/iobuf.h:16:
In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/details/io_byte_iterator.h:14:
In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/details/io_fragment.h:15:
In file included from bazel-out/k8-fastbuild/bin/src/v/base/_virtual_includes/base/base/vassert.h:17:
external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:71:9: error: call to immediate function 'crash_tracker::(anonymous namespace)::record_backtrace(crash_description &)::(anonymous
 class)::operator()' is not a constant expression
   71 |         func(decorate(ip - 1));
      |         ^
src/v/crash_tracker/recorder.cc:173:9: note: in instantiation of function template specialization 'seastar::backtrace<(lambda at src/v/crash_tracker/recorder.cc:173:19)>' requested here
  173 |     ss::backtrace([&pos, &printer](ss::frame f) {
      |         ^
external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:71:23: note: read of non-const variable 'ip' is not allowed in a constant expression
   71 |         func(decorate(ip - 1));
      |                       ^
external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:70:14: note: declared here
   70 |         auto ip = reinterpret_cast<uintptr_t>(buffer[i]);
      |              ^
1 error generated.
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 12.483s, Critical Path: 11.68s
INFO: 157 processes: 42 internal, 115 linux-sandbox.
ERROR: Build did NOT complete successfully
```

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--95093d81-3b5a-478b-8817-ecfacba27a74-->
<!--TEST RESULTS-->
<!--532f5133-e9d6-40d3-a36d-cea23e72d83d-->
<!--TEST RESULTS-->
<!--50660bd7-dd50-4c69-ba7b-1de951edef1e-->
<!--TEST RESULTS-->
<!--eb49d9e9-2c7a-437c-919e-861bdddca255-->
<!--TEST RESULTS-->
<!--84994881-4075-455f-929f-8c6caf4277ab-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62636">build#62636</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacc-4ef6-a4fb-a583f144a6f3 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacc-4ef6-a4fb-a583f144a6f3 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacc-4ef6-a4fb-a583f144a6f3 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacd-4368-9bb9-9f1fdadc5dbd | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eace-410d-b462-b3e784d21d5a | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacc-4ef6-a4fb-a583f144a6f3 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eace-410d-b462-b3e784d21d5a | FLAKY | 1/4 |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacb-45d7-a089-e1083f3667d7 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62636#0195646c-eacc-4ef6-a4fb-a583f144a6f3 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62786">build#62786</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62786#019571b0-ca6e-4423-95c6-5dd99487199b | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/62786#019571ad-a0fe-41ee-9ab2-daa60246ad15 | FLAKY | 1/2 |
| rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.topic_delete_installed_snapshots_test | ducktape | https://buildkite.com/redpanda/redpanda/builds/62786#019571b0-ca76-42fc-84fe-68316ce938ab | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62891">build#62891</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62891#019583cb-6d64-4783-80b4-a08306e01fa8 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62891#019583cb-6d64-4783-80b4-a08306e01fa8 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62915">build#62915</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585cf-6d24-4d8d-8b26-e1b041cd9fb2 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585cf-6d25-461f-9280-fc10a8977b14 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585cf-6d23-44fb-bb8e-520617d62a58 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585ea-4477-4d4b-a5cb-1307d9bf4c82 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585cf-6d25-461f-9280-fc10a8977b14 | FLAKY | 1/6 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585ea-4476-493a-927a-efe43972b049 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585ea-4476-4af2-a543-9bbe168a4587 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62915#019585ea-4476-493a-927a-efe43972b049 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63018">build#63018</a></summary>


| test_id                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63018#01958d6f-7d22-46f3-9de0-ddf658e5001a | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63018#01958d6f-7d25-4300-a0f3-6c086b626d1b | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented:
> Hrm currently stuck on this and I don't understand the error...
> 
> ```
> ERROR: /home/rockwood/code/redpanda/src/v/crash_tracker/BUILD:3:20: Compiling src/v/crash_tracker/recorder.cc failed: (Exit 1): cc_wrapper.sh failed: error executing CppCompile command (from target //src/v/cra
> sh_tracker:crash_tracker) external/toolchains_llvm~~llvm~llvm_20_toolchain/bin/cc_wrapper.sh -U_FORTIFY_SOURCE '--target=x86_64-unknown-linux-gnu' '-march=westmere' -U_FORTIFY_SOURCE -fstack-protector -fno-omi
> t-frame-pointer ... (remaining 717 arguments skipped)
> 
> Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
> In file included from src/v/crash_tracker/recorder.cc:12:
> In file included from bazel-out/k8-fastbuild/bin/src/v/crash_tracker/_virtual_includes/crash_tracker/crash_tracker/recorder.h:15:
> In file included from bazel-out/k8-fastbuild/bin/src/v/crash_tracker/_virtual_includes/crash_tracker/crash_tracker/prepared_writer.h:15:
> In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/iobuf.h:16:
> In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/details/io_byte_iterator.h:14:
> In file included from bazel-out/k8-fastbuild/bin/src/v/bytes/_virtual_includes/iobuf/bytes/details/io_fragment.h:15:
> In file included from bazel-out/k8-fastbuild/bin/src/v/base/_virtual_includes/base/base/vassert.h:17:
> external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:71:9: error: call to immediate function 'crash_tracker::(anonymous namespace)::record_backtrace(crash_description &)::(anonymous
>  class)::operator()' is not a constant expression
>    71 |         func(decorate(ip - 1));
>       |         ^
> src/v/crash_tracker/recorder.cc:173:9: note: in instantiation of function template specialization 'seastar::backtrace<(lambda at src/v/crash_tracker/recorder.cc:173:19)>' requested here
>   173 |     ss::backtrace([&pos, &printer](ss::frame f) {
>       |         ^
> external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:71:23: note: read of non-const variable 'ip' is not allowed in a constant expression
>    71 |         func(decorate(ip - 1));
>       |                       ^
> external/_main~non_module_dependencies~seastar/include/seastar/util/backtrace.hh:70:14: note: declared here
>    70 |         auto ip = reinterpret_cast<uintptr_t>(buffer[i]);
>       |              ^
> 1 error generated.
> Use --verbose_failures to see the command lines of failed build steps.
> INFO: Elapsed time: 12.483s, Critical Path: 11.68s
> INFO: 157 processes: 42 internal, 115 linux-sandbox.
> ERROR: Build did NOT complete successfully
> ```

I'm not sure why, yet, but the problematic line is:

> printer(FMT_STRING("{:#x}"), f.addr);

for example `FMT_STRING("{}")` compiles ok. From what I can tell "something" can't see the const-ness of the string literal. cc @pgellert

pgellert commented:
I couldn't figure out exactly why clang is complaining about that, but I suspect it is confused trying to deduce the constexpr-ness of the parameters of the lambda function. I put up a fix that refactors this code in a way that compiles with clang-20: https://github.com/redpanda-data/redpanda/pull/25265

pgellert commented:
Sorry, merging https://github.com/redpanda-data/redpanda/pull/25265 auto-closed this as well. I've reopened it.

rockwotj committed:
bazel: switch llvm toolchain to use vtools scripts

There are still sanitizer issues in the configuration I was using, so
switch to something that is basically the same as our vtools build.

We can overtime improve that build, but it's important to get something
working first.

rockwotj committed:
pp: fixes for libc++ 19

rockwotj committed:
bazel/toolchain: enable LTO builds of clang

rockwotj committed:
bazel/toolchain: add clang 19

Add clang 19 support, as well as upgrading the debian package to build
using clang instead of GCC. The GCC version in debian:bullseye is too
old to support the features needed in clang-19. Also I think generally
clang is used to build clang these days so we don't get compiler
warnings etc.

rockwotj commented:
Adding the enterprise team, since a lot of the tweaks needed for clang-19 are in layers owned by them.

rockwotj commented:
Okay this is ready for review 🎉

michael-redpanda reviewed (dismissed):
looks legit

rockwotj committed:
bazel: make rust crate splicing faster

We only need two platforms not the 8 that are there by default. This
makes splicing way faster. Found via bazel slack.

rockwotj commented:
Force push: rebase to fix conflicts and also update rust rules to speed up splicing

IoannisRP commented on src/v/security/config_bsl.cc:
```diff
@@ -24,16 +24,16 @@ static constexpr const char* const rule_pattern_splitter{
 
 bool regex_search(
   std::string_view msg, std::cmatch& match, const std::regex& regex) {
     return std::regex_search(
-      msg.begin(),
-      msg.end(),
+      msg.data(),
+      msg.data() + msg.size(),
```
This set of changes are to avoid dangling references to the `string_view` object, right?

rockwotj replied:
no begin for many types is not a raw pointer anymore, but a wrapper around the pointer. These APIs take pointers, so you need to change everything to data/size instead of begin/end

IoannisRP reviewed (commented):
lgtm

StephanDollberg commented:
> using and old debian container and the exact same cmake options as the vtools build

Did you bisect which of these two broke the sanitizers?

rockwotj commented:
working on it slowly, compiling clang takes forever

StephanDollberg commented:
No rush really, mostly out of interest. Whether it's something more boring like library version  mismatch with the host or some more interesting LLVM setting.

rockwotj commented:
I have a feeling it's one of these settings. The binaries we built before only dynamically linked glibc, now we dynamically link a lot more and everything works fine...


One thing that might be interesting too is that we mostly built with llvm (I didn't override as many cmake tool settings as I am now to fully use clang). I am pretty sure just using gcc works just fine, so maybe some subtle issue with mixing toolchains?

rockwotj commented:
ping, would love to get this in so I can enabled sanitizers in CI for the bazel build :)
